src/sentry/workflow_engine/processors/workflow.py

```diff
@@ -1,24 +1,91 @@
+from sentry.workflow_engine.processors.data_condition_group import evaluate_condition_group
 from sentry.workflow_engine.processors.detector import get_detector_by_event
 from sentry.workflow_engine.types import WorkflowJob
 
 logger = logging.getLogger(__name__)
 
+WORKFLOW_ENGINE_PROJECT_ID_BUFFER_LIST_KEY = "workflow_engine_project_id_buffer_list"
 
-def evaluate_workflow_triggers(workflows: set[Workflow], job: WorkflowJob) -> set[Workflow]:
+
+def get_data_condition_groups_to_fire(workflows: set[Workflow], job) -> dict[int, list[int]]:
+    workflow_action_groups: dict[int, list[int]] = defaultdict(list)
+
+    workflow_ids = {workflow.id for workflow in workflows}
+
+    workflow_dcgs = WorkflowDataConditionGroup.objects.filter(
+        workflow_id__in=workflow_ids
+    ).select_related("condition_group", "workflow")
+
+    for workflow_dcg in workflow_dcgs:
+        action_condition = workflow_dcg.condition_group
+        evaluation, result = evaluate_condition_group(action_condition, job)
+
+        if evaluation:
+            workflow_action_groups[workflow_dcg.workflow_id].append(action_condition.id)
+
+    return workflow_action_groups
+
+
+def enqueue_workflows(
+    workflows: set[Workflow],
+    job: WorkflowJob,
+) -> None:
+    event = job["event"]
+    project_id = event.group.project.id
+    workflow_action_groups = get_data_condition_groups_to_fire(workflows, job)
+
+    for workflow in workflows:
+        if random.random() < 0.01:
+            logger.info(
+                "process_workflows.workflow_enqueued",
+                extra={"workflow": workflow.id, "group": event.group.id, "project": project_id},
+            )
+        buffer.backend.push_to_sorted_set(
+            key=WORKFLOW_ENGINE_PROJECT_ID_BUFFER_LIST_KEY, value=project_id
+        )
+
+        if_dcgs = workflow_action_groups.get(workflow.id, [])
+        if not if_dcgs:
+            continue
+
+        if_dcg_fields = ":".join(map(str, if_dcgs))
+
+        value = json.dumps({"event_id": event.event_id, "occurrence_id": event.occurrence_id})
+        buffer.backend.push_to_hash(
+            model=Workflow,
+            filters={"project": project_id},
+            field=f"{workflow.id}:{event.group.id}:{if_dcg_fields}",
+            value=value,
+        )
+        metrics.incr("delayed_workflow.group_added")
+
+
+def evaluate_workflow_triggers(
+    workflows: set[Workflow], job: WorkflowJob
+) -> tuple[set[Workflow], set[Workflow]]:
     triggered_workflows: set[Workflow] = set()
+    workflows_to_enqueue: set[Workflow] = set()
 
     for workflow in workflows:
         if workflow.evaluate_trigger_conditions(job):
             triggered_workflows.add(workflow)
+        else:
+            if get_slow_conditions(workflow):
+                # enqueue to be evaluated later
+                workflows_to_enqueue.add(workflow)
 
-    return triggered_workflows
+    return triggered_workflows, workflows_to_enqueue
```

Comment:
i don't think we should be doing filtering or anything here - this method should be a pure method to evaluate the workflow triggers and that's it; if we want to filter the workflows being evaluated we should do that before evaluating them.

let's update the code to have `process_workflows` figure out what is fast / slow conditions, then filter based on fast / slow conditions.

Reply:
we only enqueue workflows that need to have slow conditions evaluated because they don't pass after evaluating the fast conditions alone. are you saying to evaluate the workflows with slow conditions separately? some of them might be triggered immediately and some of them may have to be enqueued

Reply:
updated to only return `triggered_workflows`